modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/InboundConnectionHandler.java

```diff
@@ -0,0 +1,930 @@
+        GridNioServerWrapper nioSrvWrapper,
+        CountDownLatch ctxInitLatch,
+        boolean client,
+        Supplier<Ignite> igniteExSupplier,
+        CommunicationListener<Message> lsnr
+    ) {
+        this.log = log;
+        this.cfg = cfg;
+        this.nodeGetter = nodeGetter;
+        this.locNodeSupplier = locNodeSupplier;
+        this.stateProvider = stateProvider;
+        this.clientPool = clientPool;
+        this.commWorker = commWorker;
+        this.connectGate = connectGate;
+        this.failureProcessorSupplier = failureProcessorSupplier;
+        this.attributeNames = attributeNames;
+        this.metricsLsnr = metricsLsnr;
+        this.nioSrvWrapper = nioSrvWrapper;
+        this.ctxInitLatch = ctxInitLatch;
+        this.client = client;
+        this.igniteExSupplier = igniteExSupplier;
+        this.lsnr = lsnr;
+    }
+
+    /**
+     * @param nioSrvWrapper Nio server wrapper.
+     */
+    public void setNioSrvWrapper(GridNioServerWrapper nioSrvWrapper) {
+        this.nioSrvWrapper = nioSrvWrapper;
+    }
+
+    /**
+     * @param pool Pool.
+     */
+    public void setClientPool(ConnectionClientPool pool) {
+        this.clientPool = pool;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onSessionWriteTimeout(GridNioSession ses) {
+        LT.warn(log, "Communication SPI session write timed out (consider increasing " +
+            "'socketWriteTimeout' " + "configuration property) [remoteAddr=" + ses.remoteAddress() +
+            ", writeTimeout=" + cfg.socketWriteTimeout() + ']');
+
+        if (log.isDebugEnabled())
+            log.debug("Closing communication SPI session on write timeout [remoteAddr=" + ses.remoteAddress() +
+                ", writeTimeout=" + cfg.socketWriteTimeout() + ']');
+
+        ses.close();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onConnected(GridNioSession ses) {
+        if (ses.accepted()) {
+            if (log.isInfoEnabled())
+                log.info("Accepted incoming communication connection [locAddr=" + ses.localAddress() +
+                    ", rmtAddr=" + ses.remoteAddress() + ']');
+
+            try {
+                if (client || ctxInitLatch.getCount() == 0 || !stateProvider.isHandshakeWaitSupported()) {
+                    if (log.isDebugEnabled())
+                        log.debug("Sending local node ID to newly accepted session: " + ses);
+
+                    ses.sendNoFuture(stateProvider.nodeIdMessage(), null);
+                }
+                else {
+                    if (log.isDebugEnabled())
+                        log.debug("Sending handshake wait message to newly accepted session: " + ses);
+
+                    ses.sendNoFuture(new HandshakeWaitMessage(), null);
+                }
+            }
+            catch (IgniteCheckedException e) {
+                U.error(log, "Failed to send message: " + e, e);
+            }
+        }
+        else {
+            if (log.isInfoEnabled())
```

Comment:
How can 'InboudConnectionHandler' establish outgoing connection?
Is it misbehavior or just inaccurate class name?